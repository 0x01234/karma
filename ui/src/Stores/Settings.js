import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
