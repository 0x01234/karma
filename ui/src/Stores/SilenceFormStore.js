import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
