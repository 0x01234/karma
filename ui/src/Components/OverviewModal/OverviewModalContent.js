import React, { Component } from "react";
import PropTypes from "prop-types";

import { observer } from "mobx-react";

import { AlertStore } from "Stores/AlertStore";
import { LabelWithPercent } from "Components/Labels/LabelWithPercent";

const LabelsTable = observer(({ alertStore }) => (
  <table
    className="table table-borderless top-labels"
    style={{ tableLayout: "fixed" }}
  >
    <tbody className="mw-100">
      {alertStore.data.counters.map(nameStats => (
        <tr key={nameStats.name}>
          <td width="25%" className="text-nowrap mw-100 p-1">
            <span className="badge badge-light components-label mx-0 my-1 pl-0 text-left">
              <span className="bg-primary text-white mr-1 px-1 components-labelWithPercent-percent">
                {nameStats.hits}
              </span>
              {nameStats.name}
            </span>
          </td>
          <td width="75%" className="mw-100 p-1">
            {nameStats.values.slice(0, 9).map((valueStats, i, array) => (
              <LabelWithPercent
                key={valueStats.value}
                name={nameStats.name}
                value={valueStats.value}
                hits={valueStats.hits}
                percent={valueStats.percent}
                offset={array
                  .slice(0, i)
                  .map(ns => ns.percent)
                  .reduce((a, b) => a + b, 0)}
              />
            ))}
          </td>
        </tr>
      ))}
    </tbody>
  </table>
));

const NothingToShow = () => (
  <div className="jumbotron bg-white">
    <h1 className="display-5 text-secondary text-center">
      No labels to display
    </h1>
  </div>
);

const OverviewModalContent = observer(
  class OverviewModalContent extends Component {
    static propTypes = {
      alertStore: PropTypes.instanceOf(AlertStore).isRequired,
      onHide: PropTypes.func.isRequired
    };

    render() {
      const { alertStore, onHide } = this.props;

      return (
        <React.Fragment>
          <div className="modal-header">
            <h5 className="modal-title">Overview</h5>
            <button type="button" className="close" onClick={onHide}>
              <span className="align-middle">&times;</span>
            </button>
          </div>
          <div className="modal-body">
            {alertStore.data.counters.length === 0 ? (
              <NothingToShow />
            ) : (
              <LabelsTable alertStore={alertStore} />
            )}
          </div>
        </React.Fragment>
      );
    }
  }
);

export { OverviewModalContent };
