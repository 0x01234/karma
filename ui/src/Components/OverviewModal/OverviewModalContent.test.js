import React from "react";

import { Provider } from "mobx-react";

import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { AlertStore, NewUnappliedFilter } from "Stores/AlertStore";
import { OverviewModalContent } from "./OverviewModalContent";

let alertStore;
const onHide = jest.fn();

beforeEach(() => {
  alertStore = new AlertStore([]);
  onHide.mockClear();
});

afterEach(() => {
  jest.restoreAllMocks();
});

describe("<OverviewModalContent />", () => {
  it("matches snapshot with labels to show", () => {
    alertStore.filters.values = [
      NewUnappliedFilter("abc=xyz"),
      NewUnappliedFilter("foo=bar")
    ];
    alertStore.data.counters = [
      {
        name: "foo",
        hits: 16,
        values: [
          { value: "bar1", raw: "foo=bar1", hits: 8, percent: 50, offset: 0 },
          { value: "bar2", raw: "foo=bar2", hits: 4, percent: 25, offset: 50 },
          { value: "bar3", raw: "foo=bar3", hits: 4, percent: 25, offset: 75 }
        ]
      },
      {
        name: "alertname",
        hits: 5,
        values: [
          {
            value: "Host_Down",
            raw: "alertname=Host_Down",
            hits: 5,
            percent: 100,
            offset: 0
          }
        ]
      }
    ];

    // we have multiple fragments and enzyme only renders the first one
    // in html() and text(), debug() would work but it's noisy
    // https://github.com/airbnb/enzyme/issues/1213
    const tree = mount(
      <span>
        <Provider alertStore={alertStore}>
          <OverviewModalContent alertStore={alertStore} onHide={onHide} />
        </Provider>
      </span>
    );
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("matches snapshot with no labels to show", () => {
    alertStore.data.counters = [];

    // we have multiple fragments and enzyme only renders the first one
    // in html() and text(), debug() would work but it's noisy
    // https://github.com/airbnb/enzyme/issues/1213
    const tree = mount(
      <span>
        <Provider alertStore={alertStore}>
          <OverviewModalContent alertStore={alertStore} onHide={onHide} />
        </Provider>
      </span>
    );
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });
});
