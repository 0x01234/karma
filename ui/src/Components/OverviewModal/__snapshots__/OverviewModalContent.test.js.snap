// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OverviewModalContent /> matches snapshot with labels to show 1`] = `
"
<span>
  <div class=\\"modal-header\\">
    <h5 class=\\"modal-title\\">
      Overview
    </h5>
    <button type=\\"button\\"
            class=\\"close\\"
    >
      <span class=\\"align-middle\\">
        ×
      </span>
    </button>
  </div>
  <div class=\\"modal-body\\">
    <table class=\\"table table-borderless top-labels\\"
           style=\\"table-layout: fixed;\\"
    >
      <tbody class=\\"mw-100\\">
        <tr>
          <td width=\\"25%\\"
              class=\\"text-nowrap mw-100 p-1\\"
          >
            <span class=\\"badge badge-light components-label mx-0 my-1 pl-0 text-left\\">
              <span class=\\"bg-primary text-white mr-1 px-1 components-labelWithPercent-percent\\">
                16
              </span>
              foo
            </span>
          </td>
          <td width=\\"75%\\"
              class=\\"mw-100 p-1\\"
          >
            <div class
                 style=\\"display: inline-block; max-width: 100%;\\"
                 data-tooltipped
                 aria-describedby=\\"tippy-tooltip-1\\"
                 data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
            >
              <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover mb-0 pl-0 text-left\\">
                <span class=\\"mr-1 px-1 bg-primary text-white components-labelWithPercent-percent\\">
                  8
                </span>
                <span>
                  <span class=\\"components-label-name\\">
                    foo:
                  </span>
                  <span class=\\"components-label-value\\">
                    bar1
                  </span>
                </span>
              </span>
              <div class=\\"progress components-labelWithPercent-progress mr-1\\">
                <div class=\\"progress-bar bg-warning\\"
                     role=\\"progressbar\\"
                     style=\\"width: 50%;\\"
                     aria-valuenow=\\"50\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
              </div>
            </div>
            <div class
                 style=\\"display: inline-block; max-width: 100%;\\"
                 data-tooltipped
                 aria-describedby=\\"tippy-tooltip-2\\"
                 data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
            >
              <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover mb-0 pl-0 text-left\\">
                <span class=\\"mr-1 px-1 bg-primary text-white components-labelWithPercent-percent\\">
                  4
                </span>
                <span>
                  <span class=\\"components-label-name\\">
                    foo:
                  </span>
                  <span class=\\"components-label-value\\">
                    bar2
                  </span>
                </span>
              </span>
              <div class=\\"progress components-labelWithPercent-progress mr-1\\">
                <div class=\\"progress-bar bg-transparent\\"
                     role=\\"progressbar\\"
                     style=\\"width: 50%;\\"
                     aria-valuenow=\\"50\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
                <div class=\\"progress-bar bg-success\\"
                     role=\\"progressbar\\"
                     style=\\"width: 25%;\\"
                     aria-valuenow=\\"25\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
              </div>
            </div>
            <div class
                 style=\\"display: inline-block; max-width: 100%;\\"
                 data-tooltipped
                 aria-describedby=\\"tippy-tooltip-3\\"
                 data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
            >
              <span class=\\"components-label badge badge-warning components-label-dark components-label-with-hover mb-0 pl-0 text-left\\">
                <span class=\\"mr-1 px-1 bg-primary text-white components-labelWithPercent-percent\\">
                  4
                </span>
                <span>
                  <span class=\\"components-label-name\\">
                    foo:
                  </span>
                  <span class=\\"components-label-value\\">
                    bar3
                  </span>
                </span>
              </span>
              <div class=\\"progress components-labelWithPercent-progress mr-1\\">
                <div class=\\"progress-bar bg-transparent\\"
                     role=\\"progressbar\\"
                     style=\\"width: 75%;\\"
                     aria-valuenow=\\"75\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
                <div class=\\"progress-bar bg-success\\"
                     role=\\"progressbar\\"
                     style=\\"width: 25%;\\"
                     aria-valuenow=\\"25\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
              </div>
            </div>
          </td>
        </tr>
        <tr>
          <td width=\\"25%\\"
              class=\\"text-nowrap mw-100 p-1\\"
          >
            <span class=\\"badge badge-light components-label mx-0 my-1 pl-0 text-left\\">
              <span class=\\"bg-primary text-white mr-1 px-1 components-labelWithPercent-percent\\">
                5
              </span>
              alertname
            </span>
          </td>
          <td width=\\"75%\\"
              class=\\"mw-100 p-1\\"
          >
            <div class
                 style=\\"display: inline-block; max-width: 100%;\\"
                 data-tooltipped
                 aria-describedby=\\"tippy-tooltip-4\\"
                 data-original-title=\\"Click to only show alerts with this label or Alt+Click to hide them\\"
            >
              <span class=\\"components-label badge badge-dark components-label-dark components-label-with-hover mb-0 pl-0 text-left\\">
                <span class=\\"mr-1 px-1 bg-primary text-white components-labelWithPercent-percent\\">
                  5
                </span>
                <span>
                  <span class=\\"components-label-name\\">
                    alertname:
                  </span>
                  <span class=\\"components-label-value\\">
                    Host_Down
                  </span>
                </span>
              </span>
              <div class=\\"progress components-labelWithPercent-progress mr-1\\">
                <div class=\\"progress-bar bg-danger\\"
                     role=\\"progressbar\\"
                     style=\\"width: 100%;\\"
                     aria-valuenow=\\"100\\"
                     aria-valuemin=\\"0\\"
                     aria-valuemax=\\"100\\"
                >
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</span>
"
`;

exports[`<OverviewModalContent /> matches snapshot with no labels to show 1`] = `
"
<span>
  <div class=\\"modal-header\\">
    <h5 class=\\"modal-title\\">
      Overview
    </h5>
    <button type=\\"button\\"
            class=\\"close\\"
    >
      <span class=\\"align-middle\\">
        ×
      </span>
    </button>
  </div>
  <div class=\\"modal-body\\">
    <div class=\\"jumbotron bg-white\\">
      <h1 class=\\"display-5 text-secondary text-center\\">
        No labels to display
      </h1>
    </div>
  </div>
</span>
"
`;
